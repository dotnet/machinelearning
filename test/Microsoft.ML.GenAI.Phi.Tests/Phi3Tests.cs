// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Text;
using System.Text.Json;
using ApprovalTests;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using FluentAssertions;
using Microsoft.ML.GenAI.Core.Extension;
using TorchSharp;
using Xunit;

namespace Microsoft.ML.GenAI.Phi.Tests;

[Collection("NoParallelization")]
public class Phi3Tests
{
    public Phi3Tests()
    {
        if (Environment.GetEnvironmentVariable("HELIX_CORRELATION_ID") != null)
        {
            Approvals.UseAssemblyLocationForApprovedFiles();
        }

        torch.set_default_device("meta");
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KInt8QuantizeShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        model.ToInt8QuantizeModule();
        var size = model.GetSizeInBytes();
        var stateDictStr = model.PeekShape();
        var sizeInGB = size / 1024 / 1024 / 1024;
        sizeInGB.Should().Be(3);
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini4KInt4QuantizeShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini4kInstruct);
        model.ToInt4QuantizeModule();
        var size = model.GetSizeInBytes();
        var stateDictStr = model.PeekShape();
        var sizeInGB = size / 1024 / 1024 / 1024;
        sizeInGB.Should().Be(2);
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium4KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Medium4kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }


    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Medium128KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Medium128kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KShapeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini128kInstruct);
        var stateDictStr = model.PeekShape();
        Approvals.Verify(stateDictStr);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void Phi3Mini128KLayerSizeTest()
    {
        var model = new Phi3ForCasualLM(Phi3Config.Phi3Mini128kInstruct);
        var size = model.GetSizeForEachDynamicLayerInBytes();
        // convert size to MB
        var sizeInMB = size.ToDictionary(x => x.Key, x => x.Value / 1024 / 1024);

        var json = JsonSerializer.Serialize(sizeInMB, new JsonSerializerOptions { WriteIndented = true });
        Approvals.Verify(json);
    }

    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("Approvals")]
    public void TokenizerTest()
    {
        var modelWeightFolder = Path.Join("Llama");
        var modelPath = Path.Join(modelWeightFolder, "tokenizer.model");
        var tokenizer = Phi3TokenizerHelper.FromPretrained(modelPath);
        tokenizer.BeginningOfSentenceId.Should().Be(1);
        tokenizer.EndOfSentenceId.Should().Be(2);

        // test <|end|>
        var endIds = tokenizer.EncodeToIds("<|end|>", addBeginningOfSentence: false, addEndOfSentence: false, considerPreTokenization: false, considerNormalization: false);
        endIds.Should().BeEquivalentTo(new int[] { 32007 });

        var messages = new string[]
        {
            "Can you provide ways to eat combinations of bananas and dragonfruits?",
            "Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.",
            "What about solving an 2x + 3 = 7 equation?",
            "\nCount to 3\n",
            "<|user|>",
            "<|end|>",
            "<|assistant|>",
            "<|user|>\nCount to 3<|end|>\n<|assistant|>",
        };
        var sb = new StringBuilder();
        foreach (var message in messages)
        {
            var tokenizeIds = tokenizer.EncodeToIds(message, true, false, considerPreTokenization: true);
            var decodeToString = tokenizer.Decode(tokenizeIds, considerSpecialTokens: true);
            sb.AppendLine(decodeToString);
            var tokenizedStr = string.Join(", ", tokenizeIds.Select(x => x.ToString()));

            sb.AppendLine(tokenizedStr);
        }
        Approvals.Verify(sb.ToString());
    }
}
