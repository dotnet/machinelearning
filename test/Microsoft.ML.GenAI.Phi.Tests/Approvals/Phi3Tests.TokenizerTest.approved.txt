<s>Can you provide ways to eat combinations of bananas and dragonfruits?
1, 1815, 366, 3867, 5837, 304, 17545, 18240, 310, 9892, 16397, 322, 8338, 265, 29888, 21211, 29973
<s>Sure! Here are some ways to eat bananas and dragonfruits together: 1. Banana and dragonfruit smoothie: Blend bananas and dragonfruits together with some milk and honey. 2. Banana and dragonfruit salad: Mix sliced bananas and dragonfruits together with some lemon juice and honey.
1, 18585, 29991, 2266, 526, 777, 5837, 304, 17545, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 29901, 29871, 29896, 29889, 10765, 1648, 322, 8338, 265, 29888, 9216, 10597, 347, 29901, 3164, 355, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 411, 777, 27274, 322, 298, 4992, 29889, 29871, 29906, 29889, 10765, 1648, 322, 8338, 265, 29888, 9216, 4497, 328, 29901, 23478, 269, 506, 287, 9892, 16397, 322, 8338, 265, 29888, 21211, 4208, 411, 777, 454, 3712, 3623, 625, 322, 298, 4992, 29889
<s>What about solving an 2x + 3 = 7 equation?
1, 1724, 1048, 17069, 385, 29871, 29906, 29916, 718, 29871, 29941, 353, 29871, 29955, 6306, 29973
<s>
Count to 3

1, 29871, 13, 3981, 304, 29871, 29941, 13
<s><|user|>
1, 32010
<s><|end|>
1, 32007
<s><|assistant|>
1, 32001
<s><|user|>
Count to 3<|end|>
<|assistant|>
1, 32010, 29871, 13, 3981, 304, 29871, 29941, 32007, 13, 32001
